infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftClientFactory.java

```diff
@@ -65,22 +71,56 @@ public OpenShiftClientFactory(
     if (doTrustCerts != null) {
       configBuilder.withTrustCerts(doTrustCerts);
     }
-    this.client = new UnclosableOpenShiftClient(configBuilder.build());
+
+    Config theConfig = configBuilder.build();
+    return theConfig;
+  }
+
+  private OpenShiftClient createOC(Config config) throws InfrastructureException {
+    OkHttpClient clientHttpClient =
+        getHttpClient().newBuilder().authenticator(Authenticator.NONE).build();
+    OkHttpClient.Builder builder = clientHttpClient.newBuilder();
+    builder.interceptors().clear();
+    clientHttpClient =
+        builder
+            .addInterceptor(
+                new OpenShiftOAuthInterceptor(clientHttpClient, OpenShiftConfig.wrap(config)))
+            .build();
+
+    return new UnclosableOpenShiftClient(clientHttpClient, config);
   }
 
   /**
    * Creates instance of {@link OpenShiftClient}.
    *
    * @throws InfrastructureException if any error occurs on client instance creation.
    */
-  public OpenShiftClient create() throws InfrastructureException {
-    return client;
+  public OpenShiftClient createOC(String workspaceId) throws InfrastructureException {
```

Comment:
Can you elaborate why did you add new method instead of overring parent's method?

Reply:
In some use-cases (mainly the use-cases involving `Exec` or `Watch` commands with underlying long-runing web-socket connections), the `OpenshiftClient` introduces connection leaks as soon as we create / use / manage several `OpenshiftClient` instances. This seems to be due to the `OpenShiftOAuthInterceptor`, and the structure of the `OpenshiftClient` doesn't allow us overriding this error-prone behavior.
The previous fix for this in Che6 was to use a single `OpenshiftClient` instance for all the calls, which is obviously not possible in multi-tenant scenarios.
 
However the good news is that for such use-cases (`Exec` and `Watch`), the `KubernetesClient` is fully sufficient. In fact the `KubernetesClient` is sufficient for most of the API calls in the Openshift insfastructure (which now derives from the kubernetes infrastructure BTW).

So the idea of this change is that, even in the Openshift infrastructure, we use the `KubernetesClient` in all the places where it is possible, and only use the `OpenshiftClient` for Openshift-specific API calls: routes and projects.

This is the way it has been fixed in the Che5-based OSIO currently running in production.

Reply:
I thought that caching of OkHttpClient resolves an issue with connection leaks. Doesn't it?

If not then please consider adding java doc that using OpenShift client may lead to connection leaks and it is better to avoid using it if Kubernetes client is able to do the same action.

But it looks like a workaround it's better to register an issue for fabric8 OS Client.

Reply:
> I thought that caching of `OkHttpClient` resolves an issue with connection leaks. Doesn't it?

That's not precisely the case. In order to fix connection leaks, Che 6 was initially caching the `OpenshiftClient` instance itself in a global way. But now we cannot cache it anymore since multi-tenancy involves using different `OpenshiftClient` instances for starting workspaces for different users.

> If not then please consider adding java doc that using OpenShift client may lead to connection leaks and it is better to avoid using it if Kubernetes client is able to do the same action.

Yes, that's something I will do

> But it looks like a workaround it's better to register an issue for fabric8 OS Client.

Sure, we can open an issue on the OS client, but once again, I'm not sure we want to wait for this bug / limitation to be clearly analyzed / fixed on the OS client side, a new OS client release issued, and a new CQ approved before being able to run Che6 on OSIO ...